v2/gcs-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/processing/handler/GCSToSourceStreamingHandler.java

```diff
@@ -89,6 +112,56 @@ public static String process(ProcessingContext taskContext, SpannerDao spannerDa
     return fileProcessedStartInterval;
   }
 
+  public static void writeFilteredEventsToGcs(
+      ProcessingContext taskContext,
+      Storage storage,
+      List<TrimmedShardedDataChangeRecord> filteredEvents) {
+    Matcher matcher = Pattern.compile(GCS_INPUT_DIRECTORY_REGEX).matcher(taskContext.getGCSPath());
+
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException(
+          "--gcsInputDirectoryPath is expected in a format matching "
+              + "the following regular expression: "
+              + GCS_INPUT_DIRECTORY_REGEX);
+    }
+    String bucket = matcher.group(1);
+    String path = matcher.group(2);
+    if (!path.endsWith("/")) {
+      path += "/";
+    }
+
+    String fileStartTime = taskContext.getStartTimestamp();
+    com.google.cloud.Timestamp startTs = com.google.cloud.Timestamp.parseTimestamp(fileStartTime);
+    currentIntervalStart = new org.joda.time.Instant(startTs.toSqlTimestamp());
+    org.joda.time.Instant currentIntervalEnd =
+        currentIntervalStart.plus(taskContext.getWindowDuration());
+    // File name format for filtered events is kept same as the records written to GCS by reader
+    // template
+    String gcsFileName =
+        path
+            + "filteredEvents/"
+            + taskContext.getShard().getLogicalShardId()
+            + "/"
+            + currentIntervalStart
+            + "-"
+            + currentIntervalEnd
+            + "-pane-0-last-0-of-1.txt";
+    try {
+      BlobInfo blobInfo = BlobInfo.newBuilder(bucket, gcsFileName).build();
+      storage.create(blobInfo, filteredEvents.toString().getBytes(StandardCharsets.UTF_8));
+      LOG.info(
```

Comment:
How much will this line log ? Is it per row or per batch ?

Reply:
This is per batch